### ISSUE TITLE
Error building dependencies with yarn on Arch Linux

### ISSUE THREAD

**Issue by dmp1ce** [LN-Zap/zap-desktop#420]
I just downloaded the code and tried to build the dependencies on Arch Linux with `yarn`. I'm not sure if I'm missing something obvious.

Here is my build output.
https://bpaste.net/show/d9de030db9ca

**Comment by mrfelton** [comment]
@dmp1ce did you / can you try running `npm run install-grpc` after running `yarn`?

**Comment by mrfelton** [comment]
@dmp1ce if you could test the above patch which updates the grpc package that would be great ;)

Updated code is in this fork/branch https://github.com/mrfelton/zap-desktop/tree/feat/update-grpc

**Comment by dmp1ce** [comment]
https://bpaste.net/show/d32c16d9afc0 from the branch `feat/update-grpc`. I didn't try the step `npm run install-grpc` because I assumed the `yarn` command failed. Should I just continue on with running `lnd` and running Zap?

**Comment by dmp1ce** [comment]
`npm run install-grpc` also had errors. https://bpaste.net/show/c2bf6b2ea5e1

**Comment by mrfelton** [comment]
@dmp1ce I've been working on updating various build dependencies for the project to resolve some build related issues that I was facing. It's possible that some of these updates could resolve your issues too.

Could you try with the code from this PR https://github.com/LN-Zap/zap-desktop/pull/426 (code is in this branch https://github.com/mrfelton/zap-desktop/tree/feat/update-tooling)

**Comment by dmp1ce** [comment]
I'm not sure I'm building grpc-native-core correctly but I get these errors.

https://bpaste.net/show/659c8d50e916

**Comment by mrfelton** [comment]
@dmp1ce Can you open a support request at https://github.com/grpc/grpc/issues/new please? They should be able to help you better.

**Comment by nicolasnoble** [comment]
The actual problem is here:

https://github.com/LN-Zap/zap-desktop/blob/master/app/package.json#L14

Please don't do this. `electron-rebuild` isn't doing things correctly by skipping hooks, and will break grpc every now and then, like with this issue here. Moreover, the only advantage of using `electron-rebuild` is to automatically rebuild for the installed version of electron, so you don't need to specify the target, like you're doing. So effectively, you're getting into a situation where:

-) You are creating more problems than it's worth by using `electron-rebuild`.
-) You don't get the only advantage that using `electron-rebuild` is supposed to provide.

The correct method would rather be to do `npm rebuild --target=1.8.4 --target-runtime=electron --dist-url=https://atom.io/download/electron`, which will correctly call all of the necessary hooks so it'll try grabbing our pre-built packages instead.

**Comment by nicolasnoble** [comment]
Also, unless I am mistaken, you are only using the client side of grpc. Please consider using the @grpc/grpc-js package instead, which is a pure javascript client.

**Comment by mrfelton** [comment]
@nicolasnoble The line that you point to (the `install-grpc` script) is not actually called as part of the installation process.

When you run `yarn` it runs `electron-builder install-app-deps` which causes the `postinstall`step from `app/package.json` to run, which runs the code:

```
npm rebuild --runtime=electron --target=1.8.4 --disturl=https://atom.io/download/atom-shell --build-from-source
```

See https://github.com/LN-Zap/zap-desktop/blob/master/app/package.json#L13

The `install-grpc` script which you pointed to is currently listed as an additional installation step in the advanced installation instructions (which I believe should be removed as is redundant - and problematic as you have pointed out)

See https://github.com/LN-Zap/zap-desktop/blob/master/ADVANCED.md#installing-dependencies

So unless @dmp1ce was manually running `yarn install-grpc` after running `yarn` (which I don't think he was), then I don't think this is the issue here.

@dmp1ce can you confirm which command you were running when you faced this issue please?

**Comment by dmp1ce** [comment]
This is all I did originally:

```
git clone https://github.com/LN-Zap/zap-desktop.git
cd zap-desktop
yarn
```

Afterwords I tried some different branches from @mrfelton by changing the branch with `git checkout` and then `yarn` again.

**Comment by nicolasnoble** [comment]
Alright, I did some investigation locally, there's still something wrong somewhere, @mrfelton. You say you are invoking `npm rebuild` upon doing `yarn`, but locally, I can see the extension being built instead.

I don't know how to show it exactly with yarn to be honest, but if I try installing your app, I can see the following processes being spawned on my machine:

````
/usr/lib/gcc/x86_64-linux-gnu/5/cc1plus -quiet -I ../deps/grpc -I ../deps/grpc/include -I ../deps/grpc/third_party/abseil-cpp -I /home/pixel/.electron-gyp/iojs-2.0.2/include/node -I /home/pixel/.electron-gyp/iojs-2.0.2/src -I /home/pixel/.electron-gyp/iojs-2.0.2/deps/uv/include -I /home/pixel/.electron-gyp/iojs-2.0.2/deps/v8/include -I ../deps/grpc/third_party/boringssl/include -I /home/pixel/.electron-gyp/iojs-2.0.2/deps/zlib -I /home/pixel/.electron-gyp/iojs-2.0.2/deps/cares/include -imultiarch x86_64-linux-gnu -MMD Release/obj.target/grpc/deps/grpc/src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.d [...]
```

So whatever is being run during your installation process isn't working right: the grpc extension is being rebuilt from source instead of being installed from the prebuilt binaries we are shipping. Unfortunately, I don't know yarn enough to get any useful logging output, but I suspect that `electron-builder install-app-deps` does in fact try to rebuild grpc from scratch first, which makes the `postinstall` step useless.

**Comment by mrfelton** [comment]
> the grpc extension is being rebuilt from source instead of being installed from the prebuilt binaries we are shipping

@nicolasnoble  my understanding is that the recommendation for including 3rd party binaries in an Electron app is to always force compile from code rather than trusting pre-built binaries.

Note that in our `npm rebuild` command we specifically include the `--build-from-source` flag.

As I understand it, the issue here is not that it is trying to rebuild the binaries from source, but rather that it is failing when trying to do so.

**Comment by nicolasnoble** [comment]
That recommendation is bad, really. Or at least, I can understand this stance because a lot of prebuilt packages aren't doing it right for electron, but we took a lot of care into making it right. Now, we are providing prebuilt binaries because it's excruciatingly difficult to built it, especially on platforms like Windows. I would wager that your app is fully uninstallable under Windows that is because of this. Raspberry pis is another example where the experience is painful, not because it doesn't work, but because it takes forever.

So I would highly recommend you to change this workflow for at least grpc - in case you have other native extensions in your app that is - if you want a smoother user experience.

The other bet would be to wait until Electron migrates to node's runtime 8.11.2 or above, and then you could use our pure javascript client instead.

**Comment by mrfelton** [comment]
Hey @dmp1ce - could you try with https://github.com/LN-Zap/zap-desktop/pull/445 please?

Code in https://github.com/mrfelton/zap-desktop/tree/build/use-precompiled-grpc-bindings

**Comment by mrfelton** [comment]
@dmp1ce I would suggest you remove `node_modules` and `app/node_modules` before running `yarn`

**Comment by nicolasnoble** [comment]
At least I can confirm this doesn't try to rebuild on my machine, and makes the process much faster :-)

Thanks!

**Comment by dmp1ce** [comment]
@mrfelton That worked for me. Thank you!

**Comment by Empact** [comment]
Closing as fixed by #445
